MSD - Stats Card - Create UI with mocked data

ashiagr opened the pull request:
Parent #15201 
References https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2244

The PR adds a `Today's Stats` card in the `My Site Dashboard` using the data from the `mocked_site_dashboard_cards.json` stored locally. Once the `API` changes for the new endpoint are ready, `Today's Stats` data will be consumed from the `API` and the local `JSON` will be deleted. For now, the card will display static data, navigation, and touch events will be taken up later in a different PR. 

### UI 
| Light Mode | Dark Mode | 
| :---: | :---: | 
| ![Screenshot_20220201-063920](https://user-images.githubusercontent.com/17463767/151922938-b2e7c33d-1cb7-476f-88c3-377b4633a9f9.jpg) |![Screenshot_20220201-063907](https://user-images.githubusercontent.com/17463767/151922974-8dbd6c83-1c2d-44c1-91d4-2e11a21ebc2b.jpg) | 

UI cases 
| Maximum font size | Portuguese | French  | Arabic(RTL) |
| :---: | :---: | :---: | :---: | 
| ![Screenshot_20220201-064557](https://user-images.githubusercontent.com/17463767/151923155-e758bffd-442b-42ac-a25a-0d7057437c1a.jpg)|![Screenshot_1643703152](https://user-images.githubusercontent.com/17463767/151933599-8581b136-ccb4-49ea-9192-c2d984e8325f.png)|![Screenshot_20220201-084229](https://user-images.githubusercontent.com/17463767/151923358-649f7e83-df1c-4ed3-b05c-163b32b392eb.jpg)|![Screenshot_٢٠٢٢٠٢٠١-٠٨٤٣٢٥](https://user-images.githubusercontent.com/17463767/151923359-d0b47ecc-99f9-4a0f-a25d-07fc78346ce2.jpg)|


To test:
Prerequisites
1. Launch the app
2. Select a wp com site 
3. Navigate to My Site -> App Settings -> Debug Settings
4. Ensure that the `MySiteDashboardPhase2FeatureConfig` is to set to on
5. Restart the app if needed
6. Navigate to the My Site tab
7. Notice that Today's stats card is shown.

### Known Accessibility Issue 
The order in which Voice over announces the `Today's Stats` card contents are as follows
Today's stats title -> Views title-> Visitors title -> Likes title-> Views count -> Visitors count -> Likes count

The expected order is 
Today's stats title-> Views title-> Views count -> Visitors title ->  Visitors count -> Likes title -> Likes count 

This improvement is added in the parent issue as a Todo and will be taken up in a separate PR later.


## Regression Notes
1. Potential unintended areas of impact
- Displays Today's stats card if the flag is not enabled
- Displays Today's stats card for Self hosted site 

2. What I did to test those areas of impact (or what existing automated tests I relied on)
- Manual tests

3. What automated tests I added (or what prevented me from doing so)
- Added tests in `TodaysStatsCardBuilderTest.kt`
- Added tests in `CardsSourceTest.kt`
- Added tests in `CardsBuilderTest.kt`

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Build: Update FluxC hash

Point to FluxC draft PR including initial Stats Card model.

ashiagr committed:
Feature: Add mocked json with stats card data

Temporarily added a mocked json with the stats card data till the time endpoint work is in progress.

ashiagr committed:
Feature: Pass stats card data from vm to the cards builder

ashiagr committed:
Feature: Return mocked data from the cards source

Mocked data returned only when stats feature flag is enabled.

ashiagr committed:
Refactor: Rename Stats to TodaysStats

ashiagr committed:
Refactor: Update mock success response for today's stats card

ashiagr committed:
Build: Update FluxC hash

Points to commit hash with revised today's stats model structure from the sibling FluxC draft PR handling success response for the card.

ashiagr committed:
Tests: Update existing tests with today's stats new structure

AjeshRPai committed:
Feature: Build Today's Stats card using builder

AjeshRPai committed:
Feature: Create Today's Stats VH

AjeshRPai committed:
Feature: Add exhaustive cases for Today's Stats card tracking

AjeshRPai committed:
Feature: Create Today's Stats card basic layout

AjeshRPai committed:
Style: Apply style to Today's Stats card

AjeshRPai committed:
Updates: TodaysStatsCardBuilder logic

Updates the logic by using a private function to encapsulate the logic of converting stat to UIString text by calling the Stats util.

AjeshRPai committed:
Fixes: TodaysStatsCardBuilder with formatting

AjeshRPai committed:
Adds: Tests in TodaysStatsCardBuilderTest.kt for TodaysStatsCardBuilder

Updates: CardsBuilderTest with mock for TodaysStatsCardBuilder

AjeshRPai committed:
Updates: CardsBuilderTest.kt with tests for Today's stats

AjeshRPai committed:
Removes: Unused String my_site_todays_stat_card_comments from strings.xml

ashiagr committed:
Refactor: Return mocked data on fetch

ashiagr committed:
Merge branch 'trunk' into issue/15201-stats-card-ui

# Conflicts:
#	build.gradle

AjeshRPai committed:
Updates: tools text in my_site_todays_stats_card to trigger build

tiagomar commented:
Hey @ashiagr and @AjeshRPai, the card looks good. =) 
I have only noticed a few things while playing around with fonts sizes and different languages. We might want to take a closer look but I'm not sure if it would be part of this card. 
Please find below what I have tested and a few findings.

Configuration
- [x] Feature flag ON/OFF

Login
- [x] wp.com site
- [x] Self-hosted site
- [x] Jetpack app

Stats card interactions
- [ ] ~Tapping the card redirects to Stats -> Days tab~
- [x] Tapping card’s elements [the card will be clickable as a whole]
- [x] Long tap
- [x] Multitouch
- [x] Swipe up/right/down/left
- [x] Pinch in/out

Navigation
- [ ] ~Tapping back arrow in Stats screen~
- [ ] ~Tapping device’s Back button~

Localization
- [x] English [1,743]
- [x] Portuguese [1.743]
- [x] French [1 743].
- [x] Arabic (RTL + Hindi numerals) [١٧٤٣]

Accessibility and appearance manipulations
- [x] Different fonts sizing.
- [x] Using VoiceOver.
- [x] Portrait/Landscape
- [x] Light/Dark themes.

Stats
- [ ] ~Brand new site~
- [ ] ~Typical site~
- [x] Large numbers

No data
- [x] No stats from API
- [ ] ~Connectivity~
- [ ] ~No internet – Site previously loaded~
- [ ] ~No internet – Site NOT previously loaded~

Consistency
- [ ] ~Card stats match Stats -> Days~
- [ ] ~Card translations and localized numbers match Stats -> Days~

Data refresh
- [ ] ~Switch between sites~
- [ ] ~Pull-to-refresh~

API Errors – Reference
- [ ] ~jetpack_disabled~
- [ ] ~jetpack_disconnected~
- [ ] ~unauthorized~
- [ ] ~Error for all cards~
- [ ] ~Error for Posts cards~
- [ ] ~Error for Stats card~


![image](https://user-images.githubusercontent.com/42008628/152065725-a8f1e8a4-e56c-47ce-b10c-ee985bb03e4f.png)

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/todaystat/TodaysStatsCardBuilder.kt:
```diff
@@ -0,0 +1,22 @@
+import org.wordpress.android.ui.utils.UiString.UiStringText
+import javax.inject.Inject
+
+@Suppress("TooManyFunctions")
```
Minor: We can remove this suppression.

AjeshRPai replied:
Busted for copy-pasting code 👀 😅 . Removed in 07b86db

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/todaystat/TodaysStatsCardBuilder.kt:
```diff
@@ -0,0 +1,22 @@
+        )
+    }
+
+    private fun statToUIString(stat: Int) = UiStringText(statsUtils.toFormattedString(stat))
```
Minor: Since the `i` in `UiString` is lower-cased, `I` in `...UIString` can be lower-cased as well.

AjeshRPai replied:
@ashiagr Thanks for the input, this change would improve the consistency in the naming of the function. 👍  Updated in 07b86db

ashiagr commented on WordPress/src/test/java/org/wordpress/android/ui/mysite/cards/dashboard/CardsBuilderTest.kt:
```diff
@@ -16,20 +16,38 @@ import org.wordpress.android.ui.mysite.MySiteCardAndItem.Card.DashboardCards.Das
+    }
+
+    @Test
+    fun `given no stats, when cards are built, then todays stat card is not built`() {
+        val cards = buildDashboardCards(hasTodaysStats = false)
```
Minor:  There are some inconsistencies in how we have referenced `today's stats` in the codebase. E.g.

Package name `todaystat`
Test case descriptions `todays stat` 
Constants in the test classes e.g. `STAT_VIEWS` without `TODAYS_` 

Let's try to be consistent everywhere with pluralized `stats`. Also, if we're using `todays` without an apostrophe in the naming, let's stick with that convention.

AjeshRPai replied:
@ashiagr Yes, Thank you for this suggestion. Agreed and implemented in 8cf9cf8. 👍 🚀 .

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/CardsAdapter.kt:
```diff
@@ -25,6 +27,7 @@ class CardsAdapter(
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CardViewHolder<*> {
         return when (viewType) {
             DashboardCardType.ERROR_CARD.ordinal -> ErrorCardViewHolder(parent)
+            DashboardCardType.TODAYS_STATS_CARD.ordinal -> TodaysStatsCardViewHolder(parent, uiHelpers)
```
⚠️: App crashes on switching to a different site with logs:

```
Process: org.wordpress.android.prealpha, PID: 27437
    org.apache.commons.lang3.NotImplementedException: Diff not implemented yet
        at org.wordpress.android.ui.mysite.cards.dashboard.CardsAdapter$DashboardCardsDiffUtil.areItemsTheSame(CardsAdapter.kt:70)
        at androidx.recyclerview.widget.DiffUtil.diffPartial(DiffUtil.java:233)
        at androidx.recyclerview.widget.DiffUtil.calculateDiff(DiffUtil.java:145)
        at androidx.recyclerview.widget.DiffUtil.calculateDiff(DiffUtil.java:105)
        at org.wordpress.android.ui.mysite.cards.dashboard.CardsAdapter.update(CardsAdapter.kt:52)
        at org.wordpress.android.ui.mysite.cards.dashboard.CardsViewHolder.bind(CardsViewHolder.kt:27)
        at org.wordpress.android.ui.mysite.MySiteAdapter.onBindViewHolder(MySiteAdapter.kt:75)
```

AjeshRPai replied:
@ashiagr  My bad 😓 , I missed this scenario while testing the changes. Have implemented the diff util case for `today's stats` card. Thanks for finding this bug. 🙌 . Fixed in 7f8a7d3

ashiagr replied:
No worries, actually we were using a mocked response and didn't think of switching sites which could have caught the error in the first place. Thankfully, it was a simple fix. 😄

AjeshRPai commented:
@tiagomar Thanks for testing the changes and sharing your findings. 🙌 👍 .  

It looks like the Portrait and Landscape orientation is not displayed correctly for large text and smaller devices. I had already tested the Portrait and Landscape orientation with different devices and font sizes. Wonder how I missed this. 🤔 .I will fix this issue and update with a comment in the PR.

ashiagr reviewed (commented):
Nice work @AjeshRPai on your first contribution to the `My Site Dashboard` project! I really like how the card is shaping up. 🎉 
Just consider some of my minor comments and a warning for the new changes. 🙇‍♀️

AjeshRPai committed:
Updates: TodaysStatsCardBuilder.kt with review suggestions

Removes the unnecessary suppression annotation
Updates the function name for readability

AjeshRPai committed:
Updates: The reference of Today's stats to be consistent

Updates the name of the package to todaysstats
Updates the constants and variables in TodaysStatsCardBuilderTest and CardBuilderTest
Updates the comments in styles.xml,strings.xml and dimens.xml to today's stats

AjeshRPai committed:
Fixes: Crash on switching to a different site

AjeshRPai commented:
Hey @tiagomar, after testing the font size changes, we have decided that we will take up this issue and the other accessibility improvements(voice-over) later in a different PR. I am adding it as a to-do in the parent issue(#15201). 

If there is any other issue blocking you from approving this PR, please add it as a comment. 👍 

cc: @ashiagr

tiagomar reviewed (approved):
Problems found will be addressed in following cards, all tests passed. It's all good. =)

AjeshRPai committed:
Updates: the logic of fetching Todays stats data

Updates: the CardsSource.kt to consume data from CardsSource
Updates: the tests in CardsSourceTest.kt
Removes: Mock json and json utils

AjeshRPai committed:
Adds: Tests for stats feature enabled

AjeshRPai committed:
Updates: the logic of fetching cards from local db

Updates: the fluxC version
Updates: the logic of calling getCards with cardTypes in CardsSource
Updates: tests in CardsSourceTest

ashiagr committed:
Adds: Test for build cards when stats feature enabled

ashiagr committed:
Merge branch 'trunk' into issue/15201-stats-card-ui

# Conflicts:
#	build.gradle

ashiagr committed:
Build: Update fluxc commit

ashiagr committed:
Fix existing test

ashiagr committed:
Update FluxC trunk hash

ashiagr commented:
👋 @AjeshRPai 

I've reviewed new changes to pass card types to the new dashboard cards endpoint. Changes look good to me. I've updated `FluxC` hash for the corresponding `FluxC` PR which is now merged to `trunk`, this PR is ready to be merged once I see 👍 from you.

AjeshRPai committed:
Updates: tests in CardsSourceTest

AjeshRPai committed:
Fixes: Typographic error in CardsSourceTest

ashiagr commented:
As discussed on Slack, merging it now.
